analyzer/modules/evm.go

```diff
@@ -36,6 +37,63 @@ type EVMBlockTokenData struct {
 	TokenData map[string]*EVMTokenData
 }
 
+type EVMDeterministicError struct {
+	// Note: .error is the implementation of .Error, .Unwrap etc. It is not
+	// in the Unwrap chain. Use something like
+	// `EVMDeterministicError{fmt.Errorf("...: %w", err)}` to set up an
+	// instance with `err` in the Unwrap chain.
+	error
+}
+
+func (err EVMDeterministicError) Is(target error) bool {
+	if _, ok := target.(EVMDeterministicError); ok {
+		return true
+	}
+	return false
+}
+
+// TODO: can we move this to oasis-sdk/client-sdk/go/modules/evm?
+const EVMModuleName = "evm"
+
+var (
+	ErrEVMExecutionFailed = errors.New(EVMModuleName, 2, "execution failed")
+	ErrEVMReverted        = errors.New(EVMModuleName, 8, "reverted")
+)
+
+func evmCallWithABICustom(
+	ctx context.Context,
+	source storage.RuntimeSourceStorage,
+	round uint64,
+	gasPrice []byte,
+	gasLimit uint64,
+	caller []byte,
+	contractEthAddr []byte,
+	value []byte,
+	contractABI *abi.ABI,
+	result interface{},
+	method string,
+	params ...interface{},
+) error {
+	inPacked, err := contractABI.Pack(method, params...)
+	if err != nil {
+		return fmt.Errorf("packing evm simulate call data: %w", err)
```

Comment:
This is a deterministic error too, though one could argue it happens before the EVM. But I'd still say that if you try to call a hypothetical method `balance_for_user_id(int)` with no params, that's the same class of error as calling `total_balance()` on a contract that doesn't implement it. We label the latter as `EVMDeterministicError`, so we should do it with this one too.

That said, in today's world where we tightly control the allowable ABIs, it mostly doesn't matter how we mark this error because we'll only ever call `Pack()` with a fixed and very limited set of params, so manual debugging suffices. So ultimately, your call :)

Reply:
agreed on this being deterministic. but it's different from finding out that a contract on chain doesn't support a given method. this one is not the contract's fault, and it's likely not the fault of any transient condition either. it would be a problem in the indexer's own programming, so let's not have it go down the "mark this as properly-determined-not-to-work" codepath. unfortunate about the semantic shortcoming of it really being deterministic though